### ISSUE TITLE
False Positive error regarding __bool__

### ISSUE THREAD

**Issue by Apakottur** [microsoft/pyright#9225]
**How to reproduce**
Take the following code:
```python
# test.py
from typing import NoReturn

class MyObj:
    def __bool__(self) -> NoReturn:
        raise ValueError

obj = MyObj()

x = None or obj
assert x is obj

y = 1 or obj
assert y == 1
```

Running `python test.py`  prints nothing, meaning that all the asserts are successful.
However, running `pyright test.py` gives:
```text
test.py
  test.py:12:13 - error: Invalid conditional operand of type "MyObj"
    Method __bool__ for type "MyObj" returns type "NoReturn" rather than "bool" (reportGeneralTypeIssues)
  test.py:15:10 - error: Invalid conditional operand of type "MyObj"
    Method __bool__ for type "MyObj" returns type "NoReturn" rather than "bool" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

**What is the expected behavior?**
Pyright should not report any errors.

**What is wrong?**
The `__bool__` method of `MyObj` is never called, but Pyright assumes that it is and reports an error about it.

Perhaps I am missing something?

pyright==1.1.384
python==3.12.3

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.385.
